src/main/java/com/minecolonies/coremod/colony/Colony.java

```diff
@@ -1071,38 +1034,162 @@ public void onWorldTick(@NotNull final TickEvent.WorldTickEvent event)
+
         }
 
         //  Tick Buildings
-        for (@NotNull final AbstractBuilding building : buildings.values())
-        {
+        for (@NotNull final AbstractBuilding building : buildings.values()) {
             building.onWorldTick(event);
         }
 
-        if(isDay && !world.isDaytime())
-        {
+        if (isDay && !world.isDaytime()) {
             isDay = false;
             updateOverallHappiness();
-        }
-        else if(!isDay && world.isDaytime())
-        {
+        } else if (!isDay && world.isDaytime()) {
             isDay = true;
         }
 
         updateWayPoints();
         workManager.onWorldTick(event);
     }
 
+    private void eventRaid(World raidingWorld, int level) {
+
+        if (level == 1)
+        {
+            return;
+        }
+
+        int numberOfBarbarians = level;
+        int numberOfArcherBarbarians = (int) (0.5 * level);
+        int numberOfChiefBarbarians = (int) (0.1 * level);
+
+        int hordeTotal = numberOfArcherBarbarians + numberOfBarbarians + numberOfChiefBarbarians;
+
+        if (hordeTotal > Configurations.maxBarbarianHordeSize)
+        {
+            final int maxSize = Configurations.maxBarbarianHordeSize;
+            if (hordeTotal > 40 && maxSize == 40)
+        {
+            numberOfBarbarians = 22;
+            numberOfArcherBarbarians = 16;
+            numberOfChiefBarbarians = 2;
+        }
+
+            numberOfBarbarians = hordeTotal - maxSize;
+            if (numberOfBarbarians < 0) {numberOfBarbarians = 0;} //For error handling and correct hordeTotal
+            hordeTotal = numberOfArcherBarbarians + numberOfBarbarians + numberOfChiefBarbarians;
+            if (hordeTotal > maxSize)
+            {
+                numberOfArcherBarbarians = hordeTotal - maxSize;
+                if (numberOfArcherBarbarians < 0) {numberOfArcherBarbarians = 0;} //For error handling and correct hordeTotal
+                hordeTotal = numberOfArcherBarbarians + numberOfBarbarians + numberOfChiefBarbarians;
+                if (hordeTotal > maxSize)
+                {
+                    numberOfChiefBarbarians = hordeTotal - maxSize;
+                    if (numberOfChiefBarbarians < 0) { numberOfChiefBarbarians = 0; //For error handling's sake, shouldn't ever happen, but would prefer not to crash if it does.
+                }
+                }
+            }
+        }
+
+        int x = this.getCenter().getX();
+        int y = this.getCenter().getY();
+        int z = this.getCenter().getZ();
+
+        assert raidingWorld != null; //Make sure world isn't null
+        switch(raidingWorld.rand.nextInt(7))
+        {
+            case 0:
+                x += Configurations.workingRangeTownHall+20;
+                break;
+            case 1:
+                x -= Configurations.workingRangeTownHall+20;
+                break;
+            case 2:
+                z += Configurations.workingRangeTownHall+20;
+                break;
+            case 3:
```

Comment:
You can take a look at how I calculate the round town border in the renderer for the border rendering.
It should be quite easy to calculate a random point on the border which should be less code and more flexible than this.